Add screen space shadows as renderer feature in URP

Seongdae opened the pull request:
### Checklist for PR maker
- [ ] Have you added a backport label (if needed)? For example, the `need-backport-*` label. After you backport the PR, the label changes to `backported-*`.
- [ ] Have you updated the changelog? Each package has a `CHANGELOG.md` file.
- [ ]  Have you updated or added the documentation for your PR? When you add a new feature, change a property name, or change the behavior of a feature, it's best practice to include related documentation changes in the same PR. If you do add documentation, make sure to add the relevant Graphics Docs team member as a reviewer of the PR. If you are not sure which person to add, see the [Docs team contacts sheet](https://docs.google.com/spreadsheets/d/1rgUWWgwLFEHIQ3Rz-LnK6PAKmbM49DZZ9al4hvnztOo/edit#gid=1058860420).
- [ ] Have you added a graphic test for your PR (if needed)? When you add a new feature, or discover a bug that tests don't cover, please add a graphic test.

---
### Purpose of this PR
An issue about casting shadows on transparent objects in enabling screen space shadows.
I added a new keyword for screen space shadows and change from toggling each keywords of main light shadows to enumerating keywords of those like this
 - from
``
            #pragma multi_compile _ _MAIN_LIGHT_SHADOWS
``
``
            #pragma multi_compile _ _MAIN_LIGHT_SHADOWS_CASCADE
``
 - to
``
            #pragma multi_compile _ _MAIN_LIGHT_SHADOWS _MAIN_LIGHT_SHADOWS_CASCADE _MAIN_LIGHT_SHADOWS_SCREEN
``

It will be the same shader variants as before even including the new keyword


---
### Testing status
**Yamato**: https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Fenable_ss-shadow

Any test projects to go with this to help reviewers?

---
### Comments to reviewers
Changes
 -  add a new keyword for screen space shadow in all .shader files
 1. change from toggling keywords related to main light shadows to enumerating all keywords of main light shadows
 2. choose one between main light shadows(no cascade), main light shadows cascades, and main light shadows screen.
 - add a screen space shadows as renderer feature and three render passes for this
 1. one is screen space shadows render pass
 2. second one is enable a keyword for screen space shadows before rendering opaques
 3. third one is disable a keyword for screen space shadows after rendering opaques
 - also update stripping shaders considering the new keyword for screen space shadows
 - and... fix wrong letters

The rest of jobs
- remove built-in screen space shadows pass and related variables in render data?
- keyword name is ok? about "_MAIN_LIGHT_SHADOWS"
- why view proj matrix's z axis was flipped? -- reverse sign of z in screen space shadows shader --
- add a graphic test for screen space shadows

kaychang-unity commented on com.unity.render-pipelines.universal/Shaders/Utils/ScreenSpaceShadows.shader:
```diff
@@ -63,7 +59,7 @@ Shader "Hidden/Universal Render Pipeline/ScreenSpaceShadows"
 
             float3 vpos = ComputeViewSpacePosition(input.uv.zw, deviceDepth, unity_CameraInvProjection);
-            float3 wpos = mul(unity_CameraToWorld, float4(vpos, 1)).xyz;
+            float3 wpos = mul(unity_CameraToWorld, float4(vpos.x, vpos.y, -vpos.z, 1)).xyz;
```
I am also not sure why this is needed.

Seongdae replied:
Before binding it(https://github.com/Unity-Technologies/Graphics/blob/master/com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs#L87)
don't need to reverse z-value. I compared it with the previous code, there is no those binding vars.
I don't know which one is correct in Unity(left or right handed coordinates)

phi-lira replied:
Hm, I wonder if this in an issue in the ComputeViewSpacePosition function? The z-reverse is platform specific and it should be handled in this case by our unity_CameraInvProjection. 

```
float3 ComputeViewSpacePosition(float2 positionNDC, float deviceDepth, float4x4 invProjMatrix)
{
    float4 positionCS = ComputeClipSpacePosition(positionNDC, deviceDepth);
    float4 positionVS = mul(invProjMatrix, positionCS);
    // The view space uses a right-handed coordinate system.
    positionVS.z = -positionVS.z;
    return positionVS.xyz / positionVS.w;
}
```

Quick way to test: switch to OpenGL. :)

Seongdae replied:
I tried to test. it was the same result. I need to analyze how to handle it :) thanks.

phi-lira commented on com.unity.render-pipelines.universal/Runtime/Passes/TransparentSettingsPass.cs:
```diff
@@ -31,12 +30,16 @@ public override void Execute(ScriptableRenderContext context, ref RenderingData
+                int cascadesCount = shadowData.mainLightShadowCascadesCount;
+
+                bool shouldReceiveNoCascades = m_shouldReceiveShadows && cascadesCount == 1;
+                bool shouldReceiveCascades = m_shouldReceiveShadows && cascadesCount > 1;
```
very minor suggestion to make this more readable: name receiveShadowsNoCascade and receiveShadowsCascades.

Seongdae replied:
I applied it

phi-lira commented on com.unity.render-pipelines.universal/Runtime/RendererFeatures/ScreenSpaceShadows.cs:
```diff
@@ -0,0 +1,230 @@
+            CoreUtils.Destroy(m_Material);
+        }
+
+        private bool LoadMaterial()
```
Suggestion: today we reference the screen space shadow shader in the renderer asset. (Shaders struct). Since now we have a feature for it we could reference the shader in the feature. This way, we only include the shader in the build if the feature is added to any renderer as opposed to today that we include it even when we don't use it.

Seongdae replied:
```csharp
            [Reload("Shaders/Utils/ScreenSpaceShadows.shader")]
            public Shader screenSpaceShadowPS;
```

Does you mean the shader in renderer is needed to be removed?
I followed a way in screen space ambient occlusion

phi-lira replied:
hm, if the API is public we can't remove it (should make obsolete in that case), but can remove the reference and the [Reload] attribute.

phi-lira replied:
Likewise you can add the same construct to your ScreenSpaceShadow shader so you make sure that you include the shader in the build if the feature is reference in any renderer.

Seongdae replied:
Add obsolete attribute and remove reload and reference, also remove material of screen space shadow in forward and deferred renderer

phi-lira commented on com.unity.render-pipelines.universal/Runtime/RendererFeatures/ScreenSpaceShadows.cs:
```diff
@@ -0,0 +1,230 @@
+            {
+                renderer.EnqueuePass(m_SSShadowsPass);
+                renderer.EnqueuePass(m_EnableKeywordPass);
+                renderer.EnqueuePass(m_DisableKeywordPass);
```
Suggestion to make less boiler plate code. You don't necessarily need the enable and disable the keywords by adding passes for it. You can enable the keyword in the SSShadowPass, you can disable the keyword in the SrciptableRenderer before drawing the transparent block.

Seongdae replied:
Remove a enabling keyword pass, but it doesn't seem for the position before drawing transparent blocks to be proper, so I added disabling the keyword in transparent settings pass. if so the transparent settings pass must be called in all cases of graphics settings, because the transparent settings pass can't know whether the screen space shadow was called or not.(If we can investigate if any render pass was enqueued or not, it will be useful IMO)

kaychang-unity replied:
If you disable the keyword inside the transparent pass or ScriptableRenderer, doesn't that make this ScriptableRendererFeature not self-contained anymore? Can users actually insert ScriptableRendererFeature in other ScriptableRenderers?

Seongdae replied:
Yes, I pulled out those enabling/disabling keyword pass. and in TransparentSettingsPass 'screen space shadow' keyword will be disabled. this TransparentSettingsPass was called in the case of not receiving shadows on transparent objects from 'Setup' function of that, but in my PR the transparent pass will be always called :(

Seongdae commented on com.unity.render-pipelines.universal/Shaders/Utils/ScreenSpaceShadows.shader:
```diff
@@ -52,14 +52,11 @@ Shader "Hidden/Universal Render Pipeline/ScreenSpaceShadows"
 
-            float3 vpos = ComputeViewSpacePosition(input.uv.zw, deviceDepth, unity_CameraInvProjection);
-            float3 wpos = mul(unity_CameraToWorld, float4(vpos.x, vpos.y, -vpos.z, 1)).xyz;
+            float3 wpos = ComputeWorldSpacePosition(input.uv.xy, deviceDepth, unity_MatrixInvVP);
```
Changed from double transform to direct transform

Seongdae commented on com.unity.render-pipelines.universal/Shaders/Utils/ScreenSpaceShadows.shader:
```diff
@@ -52,14 +52,11 @@ Shader "Hidden/Universal Render Pipeline/ScreenSpaceShadows"
 
-            float3 vpos = ComputeViewSpacePosition(input.uv.zw, deviceDepth, unity_CameraInvProjection);
-            float3 wpos = mul(unity_CameraToWorld, float4(vpos.x, vpos.y, -vpos.z, 1)).xyz;
+            float3 wpos = ComputeWorldSpacePosition(input.uv.xy, deviceDepth, unity_MatrixInvVP);
```
Change from double transform to direct transform

Seongdae replied:
how about this @phi-lira @kaychang-unity

phi-lira commented on com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs:
```diff
@@ -744,10 +744,6 @@ static void InitializeShadowData(UniversalRenderPipelineAsset settings, NativeAr
 
-            // We no longer use screen space shadows in URP.
-            // This change allows us to have particles & transparent objects receive shadows.
-            shadowData.requiresScreenSpaceShadowResolve = false;// shadowData.supportsMainLightShadows && supportsScreenSpaceShadows && settings.shadowCascadeOption != ShadowCascadesOption.NoCascades;
```
Is this initialized to false anywhere else? otherwise we might have garbage here.

Seongdae replied:
Let me restore it and add 'obsolete' attribute, thanks

phi-lira commented on com.unity.render-pipelines.universal/Runtime/UniversalRenderPipelineCore.cs:
```diff
@@ -186,7 +186,6 @@ public bool IsCameraProjectionMatrixFlipped()
     [MovedFrom("UnityEngine.Rendering.LWRP")] public struct ShadowData
     {
         public bool supportsMainLightShadows;
-        public bool requiresScreenSpaceShadowResolve;
```
Make obsolete instead of removing. We have to give users at least a package to have a change to upgrade.

phi-lira replied:
This should have been done before but I guess we missed it.

Seongdae replied:
Let me restore it and add 'obsolete' attribute

phi-lira commented on com.unity.render-pipelines.universal/Shaders/Utils/ScreenSpaceShadows.shader:
```diff
@@ -56,14 +52,11 @@ Shader "Hidden/Universal Render Pipeline/ScreenSpaceShadows"
-
-#if UNITY_REVERSED_Z
-            deviceDepth = 1 - deviceDepth;
+#if SHADER_API_GLCORE
```
what about gles and WebGL? UNITY_REVERSED_Z is more generic regarding z direction in all platforms that don't reverse z.

phi-lira replied:
I'm not sure if I understand it fully, but I suppose indeed the 1 - deviceDepth should be removed (handled in matrix), then if we are doing the second part to convert from 0 - 1 to opengl -1 <-> 1 than I guess we should do that for all OpenGL platforms?

Seongdae replied:
It doesn't reverse depth, it broadens the range [-1, 1] in opengl series. that's why I use 'SHADER_API_GLCORE' instead of 'UNITY_REVERSED_Z'. I need to add other opengl series.

Seongdae replied:
I checked it again. it is better to use 'UNITY_REVERSED_Z' as you said :)

phi-lira commented:
Adding @eh-unity for extra coverage, @Unity-Technologies/gfx-qa-foundation this PR is risky we need a wide pass to check we don't regress shadows, including GLES. Maybe test the universal rendering examples + boat attack?

phi-lira commented:
Adding @thomas-zeng as this might affect XR.

Seongdae commented on com.unity.render-pipelines.universal/Runtime/Passes/ScriptableRenderPass.cs:
```diff
@@ -29,7 +29,7 @@ public enum ScriptableRenderPassInput
         AfterRenderingShadows = 100,
         BeforeRenderingPrepasses = 150,
-        AfterRenderingPrePasses = 200,
+        AfterRenderingPrepasses = 200,
```
Whenever I was working on custom render passes, it bothered me. Should I restore this? @phi-lira

Seongdae replied:
Make previous event obsolete and new one as 'AfterRenderingPrePasses'

phi-lira commented:
@Seongdae we can't break public API. If you think it's worth it you can make the old event name obsolete then add new one.

Seongdae commented on com.unity.render-pipelines.universal/Runtime/Passes/TransparentSettingsPass.cs:
```diff
@@ -31,12 +31,17 @@ public override void Execute(ScriptableRenderContext context, ref RenderingData
-                CoreUtils.SetKeyword(cmd, ShaderKeywordStrings.MainLightShadowCascades, m_shouldReceiveShadows);
+                CoreUtils.SetKeyword(cmd, ShaderKeywordStrings.MainLightShadows, receiveShadowsNoCascade);
+                CoreUtils.SetKeyword(cmd, ShaderKeywordStrings.MainLightShadowsCascades, receiveShadowsCascades);
+                CoreUtils.SetKeyword(cmd, ShaderKeywordStrings.MainLightShadowsScreen, false);
```
Add comment for the 'false' and fix an issue that enables main light shadow when main light shadow was disabled

Seongdae commented on com.unity.render-pipelines.universal/Editor/ShaderGraph/Targets/UniversalTarget.cs:
```diff
@@ -716,19 +716,17 @@ static class CoreKeywordDescriptors
+            entries = new KeywordEntry[]
+            {
+                new KeywordEntry() { displayName = "Off", referenceName = "" },
+                new KeywordEntry() { displayName = "No Cascade", referenceName = "MAIN_LIGHT_SHADOWS" },
```
I wonder if I could use keywords not matched with keywords in built-in shader. I tried to follow them as possible as I can do.

the below seems to be better than those options I proposed?
_MAIN_LIGHT_SHADOWS_OFF, _MAIN_LIGHT_SHADOWS, _MAIN_LIGHT_SHAODWS_CASCADE, _MAIN_LIGHT_SHADOWS_SCREEN

eh-unity replied:
If by built-in you mean the built-in renderer or the keywords set by the engine then I guess you don't have a choice. If you mean URP standard shaders then they do share lighting.hlsl etc. so it's still probably best to be consistent.
Current keywords (in code) look good to me, consistent and in positive form. Only question is the `_` prefix as phi-lira mentioned, below.

I think I probably meant the displayName here in my previous comment. Not sure where it shows, but "No cascade" would be better described directly as Basic, Simple or Single etc. in positive form and not through cascade option as a negative.

Seongdae commented:
Recently, fixed some issues and bugs reported from @cinight 
 - Added stripping shader variants by checking screen space shadow feature
 - Can't avoid adding an additional pass for restoring keywords of main light shadow for solving transparent objects(not receiving shadows) and FPSCameraDemo

I couldn't enable or disable relevant keywords related to main light shadow in 'Transparent Setting Pass' because it is not possible to inquire pre-enqueued render pass now.

phi-lira commented on com.unity.render-pipelines.universal/Editor/ShaderGraph/Targets/UniversalTarget.cs:
```diff
@@ -686,19 +686,17 @@ static class CoreKeywordDescriptors
+            entries = new KeywordEntry[]
+            {
+                new KeywordEntry() { displayName = "Off", referenceName = "" },
+                new KeywordEntry() { displayName = "No Cascade", referenceName = "MAIN_LIGHT_SHADOWS" },
```
Does shader graph prepend the missing `_` here?

Seongdae replied:
Yes, shader graph add automatically the '_' in front of all keywords. If the missing '_' was added to the referenceName, the name would be like '__MAIN_LIGHT_SHADOWS'.

phi-lira commented on com.unity.render-pipelines.universal/Runtime/UniversalRenderPipelineCore.cs:
```diff
@@ -284,7 +285,8 @@ public struct PostProcessingData
     {
         public static readonly string MainLightShadows = "_MAIN_LIGHT_SHADOWS";
-        public static readonly string MainLightShadowCascades = "_MAIN_LIGHT_SHADOWS_CASCADE";
+        public static readonly string MainLightShadowsCascades = "_MAIN_LIGHT_SHADOWS_CASCADE";
```
This is public API breakage, unfortunately

Seongdae replied:
I just want to match name convention, it was just my thoughts. breaking public API is needed to be careful. let me restore it as before

phi-lira commented on com.unity.render-pipelines.universal/Runtime/RendererFeatures/ScreenSpaceShadows.cs:
```diff
@@ -0,0 +1,206 @@
+                {
+                    cmd.SetViewProjectionMatrices(Matrix4x4.identity, Matrix4x4.identity);
+                    cmd.DrawMesh(RenderingUtils.fullscreenMesh, Matrix4x4.identity, material);
+                    cmd.SetViewProjectionMatrices(camera.worldToCameraMatrix, camera.projectionMatrix);
```
This might cause some issues with XR. @thomas-zeng 
 We do have internal functions to set camera matrices:
 `RenderingUtils.SetViewAndProjectionMatrices(cmd, cameraData.GetViewMatrix(), cameraData.GetGPUProjectionMatrix(), false);`

Seongdae replied:
Ok, thanks. let me use it

phi-lira commented on com.unity.render-pipelines.universal/Runtime/RendererFeatures/ScreenSpaceShadows.cs:
```diff
@@ -0,0 +1,206 @@
+
+                    // then enable main light shadows with or without cascades
+                    CoreUtils.SetKeyword(cmd, ShaderKeywordStrings.MainLightShadows, receiveShadowsNoCascade);
+                    CoreUtils.SetKeyword(cmd, ShaderKeywordStrings.MainLightShadowsCascades, receiveShadowsCascades);
```
Not an issue but minor suggestion, restoring keywords could also be done in ScriptableRenderer before rendering opaques block. Othewise by default this pass will switch to Camera target and might cause a render target switch for no reason if the previous pass is not rendering to camera target.

Seongdae replied:
It restored those shadow keywords after rendering opaque block for preventing transparent objects from screen space shadow. Those processes can be in transparent setting pass, but there was a issue in UniversalRenderingExamples reported by @cinight .

That sample added renderer feature for FPS camera, this feature was injected after rendering opaque objects. but my restoring keywords was set before transparent pass. the screen space shadow would be incorrect for fps gun space. Because the project matrix was not matched between world space and fps gun space. So I can't avoid making another pass for restoring pass and injecting the pass after rendering opaque pass.

phi-lira commented on com.unity.render-pipelines.universal/Runtime/Passes/ScriptableRenderPass.cs:
```diff
@@ -28,7 +28,9 @@ public enum ScriptableRenderPassInput
         AfterRenderingShadows = 100,
-        BeforeRenderingPrepasses = 150,
+        BeforeRenderingPrePasses = 150,
+        [Obsolete("Obsolete, to match the capital from 'Prepass' to 'PrePass'")]
```
Also add [EditorBrowsable(EditorBrowsableState.Never)] so it doesn't show up in code completion/assist tooling.

Seongdae replied:
I added it

phi-lira reviewed (approved):
LGTM. It stills bother me a little bit we have not the optimal path to combine SSAO + ScreenSpace shadow resolve, but I understand why it was done like this.

eh-unity reviewed (approved):
LGTM.

phi-lira commented:
What's the status on this PR? Please add `ready-to-merge` once all reviwers approved and ABV is green.

Seongdae commented:
@ellioman @cinight Could you check this RP?

cinight commented:
There are still two active issues:
 - Artefact when using Screen Space Shadow on UniversalRenderingExamples > FPSCameraDemo scene
 - UniversalRenderingExamples > UnityOcclusionDemo scene crystal rendering black in this PR branch

will test again when issues are fixed

ellioman commented on com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:
```diff
@@ -341,7 +342,11 @@ half AdditionalLightShadow(int lightIndex, float3 positionWS, half4 shadowMask,
 
 float4 GetShadowCoord(VertexPositionInputs vertexInput)
 {
+#ifdef _MAIN_LIGHT_SHADOWS_SCREEN
```
Change to `#if defined(_MAIN_LIGHT_SHADOWS_SCREEN)` for consistency

Seongdae replied:
Ok, I will. thanks :)

ellioman reviewed (approved):
Looks good. Need to fix the conflicts.

I'd prefer that the deprecated things would be moved to the appropriate Deprecated CS/HLSL files but I will do a pass for Master (11.x.x) soon so I can take care of that.

cinight commented:
New issue found. Deferred shadow is broken in branch. 
Detail please see bug#04 in [test doc](https://docs.google.com/document/d/1oJGcHOmGzrdhvZxxd3ISuoyFajVoG0koeTIlyqI_FSk/edit#)

Seongdae commented:
@cinight I checked two issues you reported.

Seongdae commented on com.unity.render-pipelines.universal/Shaders/Utils/StencilDeferred.shader:
```diff
@@ -184,8 +184,12 @@ Shader "Hidden/Universal Render Pipeline/StencilDeferred"
-                    float4 shadowCoord = TransformWorldToShadowCoord(posWS.xyz);
+                #if defined(MAIN_LIGHT_CALCULATE_SHADOWS)
+                    #if defined(_MAIN_LIGHT_SHADOWS_SCREEN)
+                        float4 shadowCoord = float4(screen_uv, 0.0, 1.0);
```
@kaychang-unity In case of 'Screen Space Shadow', it needs screen uv to compute shadowing. I tried to use 'ComputeScreenPos(input.positionCS.xy), but it doesn't work correctly and also tried 'ComputeScreenPos(TransformWorldToHClip(posWS)), it works well.

but the second needs more computation and I think this is not good idea. So I put directly 'screen_us' for the performance.

I can't understand why that positionCS was different between forward pass and deferred pass

kaychang-unity replied:
I think using `screen_uv` should be the fastest way to sample screen space texture. It's already done for other gbuffer textures, so you should keep using `screen_uv` instead.

As for `ComputeScreenPos(input.positionCS.xy)`, I see that is it called only from vertex shader, not from fragment shader. In vertex shader, `input.positionCS.xy` will store vertex position in homogeneous space but before perspective-divide by `input.positionCS.w`. So the value after perspective-divide by `input.positionCS.w` will be in range [-1; 1]. In pixel shader, `input.positionCS.xy` stores actual pixel coordinates (not the same as vertex shader), which will be different. I presume that is why `ComputeScreenPos(input.positionCS.xy)` does not work for you in the pixel shader.

Seongdae replied:
Ahh.. in fragment shader positionCS is normalized device coordinate! Thanks.

kaychang-unity replied:
In fragment, shader positionCS.xy is pixel coordinates. For instance, if the render-target resolution is 1280x720, it will be x=[0;1280], y=[0;720] (with half pixel offset).
positionCS is normalized device coordinate in vertex shader only.

kaychang-unity commented on com.unity.render-pipelines.universal/Shaders/Utils/ScreenSpaceShadows.shader:
```diff
@@ -53,15 +49,14 @@ Shader "Hidden/Universal Render Pipeline/ScreenSpaceShadows"
+            float deviceDepth = SAMPLE_TEXTURE2D_X(_CameraDepthTexture, sampler_CameraDepthTexture, input.uv.xy).r;
+#else
+            float deviceDepth = SAMPLE_TEXTURE2D_X(_CameraDepthTexture, sampler_CameraDepthTexture, input.uv.xy).r;
+            deviceDepth = deviceDepth * 2.0 - 1.0;
```
Are you using UNITY_REVERSED_Z as a shortcut to detect non-GL platforms?

kaychang-unity replied:
Would `#if defined(SHADER_API_GLCORE) || defined(SHADER_API_GLES) || defined(SHADER_API_GLES3)` be equivalent but more correct?

Seongdae replied:
Yes, it is general to use it to detect platform instead of using platform specification defines.

Seongdae replied:
I used 'SHADER_API_GLCORE' before, but @phi-lira  suggested me to use 'UNITY_REVERSED_Z'.

kaychang-unity replied:
I see!

cinight commented:
[Bug#04] verified fixed
[Bug#05] still reproducible in latest commit

cinight commented:
Verified above issues. Found this new one.
[Bug#07] Gamma color space post processing is much stronger in branch

Seongdae commented:
It seems to be changes in post-processing in master branch, after merging master into my branch the result is the same as master now. Thanks @cinight

cinight reviewed (approved):
Issues are all fixed and feature working fine. All good now. See [test doc](https://docs.google.com/document/d/1oJGcHOmGzrdhvZxxd3ISuoyFajVoG0koeTIlyqI_FSk/edit#) for details.

phi-lira commented:
Yamato has some failures that might be caused by this PR: 
[VFX Darker rendering in some spheres](https://yamato-artifactviewer.prd.cds.internal.unity3d.com/e2c1ed3f-e003-4f6b-968f-ad54cc178c86%2FTest_report%2FUtrTestReport2%2FTestProjects%2FVisualEffectGraph_URP%2Ftest-results/TestReportV1.html)
[Shader Graph failing tests on shader keywords ](https://yamato-artifactviewer.prd.cds.internal.unity3d.com/b8343b43-916c-4c7e-ad34-2b13de878bb8%2FTest_report%2FUtrTestReport1%2FTestProjects%2FShaderGraph%2Ftest-results/TestReportV1.html)

phi-lira reviewed (approved):
LGTM. For every new feature we now add an entry on What's New page in URP.
https://docs.unity3d.com/Packages/com.unity.render-pipelines.universal@10.2/manual/whats-new/urp-whats-new.html

This PR was opened before we had the page so not blocking the PR because of that. We can add the What's new in follow up PR.

phi-lira commented on com.unity.render-pipelines.universal/Runtime/ForwardRendererData.cs:
```diff
@@ -40,7 +40,7 @@ public sealed class ShaderResources
             public Shader copyDepthPS;
 
-            [Reload("Shaders/Utils/ScreenSpaceShadows.shader")]
+            [Obsolete("Obsolete, this feature will be supported by 'RendererFeature' in the future")]
```
Could update the text to say to mention the renderer feature we added that replaces this.

Seongdae replied:
Applied it :)

phi-lira commented on com.unity.render-pipelines.universal/Runtime/Passes/ScriptableRenderPass.cs:
```diff
@@ -48,7 +49,11 @@ public enum ScriptableRenderPassInput
         /// Camera matrices and stereo rendering are already setup at this point.
         /// </summary>
-        BeforeRenderingPrepasses = 150,
+        BeforeRenderingPrePasses = 150,
+
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        [Obsolete("Obsolete, to match the capital from 'Prepass' to 'PrePass'")]
+        BeforeRenderingPrepasses = 151,
```
Giving a second look at this, we can make this better by using ` (UnityUpgradable) -> ` defined here: https://confluence.unity3d.com/display/DEV/Obsolete+API+updaters

 So we avoid adding a new event and just rename the old one.

Seongdae replied:
Applied it :)

phi-lira commented on com.unity.render-pipelines.universal/Runtime/RendererFeatures/ScreenSpaceShadows.cs:
```diff
@@ -0,0 +1,214 @@
+            }
+        }
+
+        private class RestoreShadowKeywordsPass : ScriptableRenderPass
```
You can add a call to ConfigureTarget(BuitinRenderTextureType.Current) to avoid this pass changing the render pass to camera by default.

Seongdae replied:
Applied it :)

phi-lira commented on com.unity.render-pipelines.universal/Runtime/UniversalRenderPipelineCore.cs:
```diff
@@ -196,6 +196,7 @@ public bool IsCameraProjectionMatrixFlipped()
     [MovedFrom("UnityEngine.Rendering.LWRP")] public struct ShadowData
     {
         public bool supportsMainLightShadows;
+        [Obsolete("Obsolete, this feature will be supported by 'RendererFeature' in the future")]
```
Update text to mention new ScreenSpaceShadow renderer feature.

Seongdae replied:
Applied it :)

phi-lira reviewed (commented):
Made some extra comments but they can be addressed in follow up PRs. No urgency.

phi-lira commented:
OSX Failure: Known issue. Red DoF.
 Window Standalone Failure: Not related to this PR.
 Android Failure: Investigating...

phi-lira commented:
Android Failure: Not related to this PR.
